File: fix-encoding.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Function to fix double-encoded UTF-8
function fixDoubleEncoding(text) {
  // Convert the string to a buffer treating it as Latin-1 (ISO-8859-1)
  // then decode it as UTF-8
  const buffer = Buffer.from(text, 'latin1');
  return buffer.toString('utf8');
}

// Function to recursively fix all strings in an object
function fixObject(obj) {
  if (typeof obj === 'string') {
    return fixDoubleEncoding(obj);
  } else if (Array.isArray(obj)) {
    return obj.map(fixObject);
  } else if (typeof obj === 'object' && obj !== null) {
    const fixed = {};
    for (const [key, value] of Object.entries(obj)) {
      fixed[key] = fixObject(value);
    }
    return fixed;
  }
  return obj;
}

// Files to fix
const files = ['yo.json', 'ig.json', 'ha.json'];

files.forEach(filename => {
  const filePath = path.join(__dirname, 'src', 'locales', filename);
  
  console.log(`\nProcessing ${filename}...`);
  
  try {
    // Read the corrupted file, stripping BOM if present
    let rawData = fs.readFileSync(filePath, 'utf8');
    
    // Remove BOM if present
    if (rawData.charCodeAt(0) === 0xFEFF) {
      rawData = rawData.slice(1);
    }
    
    // Parse JSON (may contain garbled UTF-8)
    const data = JSON.parse(rawData);
    
    // Fix the encoding
    const fixed = fixObject(data);
    
    // Write back with proper UTF-8 encoding (no BOM)
    fs.writeFileSync(filePath, JSON.stringify(fixed, null, 4), { encoding: 'utf8' });
    
    console.log(`✅ Fixed ${filename}`);
    
    // Show samples
    console.log(`  loading: ${fixed.common.loading}`);
    console.log(`  error: ${fixed.common.error}`);
  } catch (error) {
    console.error(`❌ Error fixing ${filename}:`, error.message);
    console.error(`  Stack:`, error.stack);
  }
});

console.log('\n✨ Encoding fix complete!');

File: fix-encoding.py
import json
import codecs

def fix_encoding(text):
    """Fix double-encoded UTF-8 strings"""
    if not isinstance(text, str):
        return text
    try:
        # Encode as latin-1 then decode as utf-8
        return text.encode('latin-1').decode('utf-8')
    except (UnicodeDecodeError, UnicodeEncodeError):
        return text

def fix_dict(obj):
    """Recursively fix all strings in a dictionary"""
    if isinstance(obj, dict):
        return {key: fix_dict(value) for key, value in obj.items()}
    elif isinstance(obj, list):
        return [fix_dict(item) for item in obj]
    elif isinstance(obj, str):
        return fix_encoding(obj)
    return obj

# Files to fix
files = ['yo.json', 'ig.json', 'ha.json']

for filename in files:
    filepath = f'src/locales/{filename}'
    print(f'\nProcessing {filename}...')
    
    try:
        # Read with UTF-8 encoding
        with open(filepath, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # Fix encoding
        fixed_data = fix_dict(data)
        
        # Write back with UTF-8 encoding, no BOM
        with open(filepath, 'w', encoding='utf-8', newline='\n') as f:
            json.dump(fixed_data, f, ensure_ascii=False, indent=4)
        
        print(f'✅ Fixed {filename}')
        
        # Show sample
        if 'common' in fixed_data and 'loading' in fixed_data['common']:
            print(f'Sample - loading: {fixed_data["common"]["loading"]}')
            print(f'Sample - error: {fixed_data["common"]["error"]}')
    
    except Exception as e:
        print(f'❌ Error fixing {filename}: {e}')

print('\n✨ Encoding fix complete!')
